### ISSUE TITLE
Build fails because N_PATH_HOME is too small

### ISSUE THREAD

**Issue by ryandesign** [espeak-ng/espeak-ng#1004]
[espeak-ng was added to MacPorts](https://ports.macports.org/port/espeak-ng) last year but [a user reported to MacPorts](https://trac.macports.org/ticket/63384) yesterday, and I confirmed on one of my machines, that espeak-ng 1.50 fails to build in a standard MacPorts install.

The build runs these commands:

```sh
ESPEAK_DATA_PATH=/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_audio_espeak-ng/espeak-ng/work/espeak-ng-1.50 src/espeak-ng --compile-intonations && \
    ESPEAK_DATA_PATH=/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_audio_espeak-ng/espeak-ng/work/espeak-ng-1.50 src/espeak-ng --compile-phonemes && \
    touch phsource/phonemes.stamp
```

And the error is:

```sh
Error processing file '/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_audio_espeak-ng/espeak-ng/work/espeak-ng-1.50//../phsource/intonation': No such file or directory.
```

Note the two slashes in `espeak-ng-1.50//../phsource`. `espeak-ng-data` should appear between those two slashes but doesn't, which causes the problem.

The code that appends "/espeak-ng-data" to the ESPEAK_DATA_PATH is in `check_data_path` in src/libespeak-ng/speech.c:

```c
snprintf(path_home, sizeof(path_home), "%s/espeak-ng-data", path);
```

`path_home` is defined earlier in that file:

```c
char path_home[N_PATH_HOME]; // this is the espeak-ng-data directory
```

`N_PATH_HOME` is defined in speech.h as:

```c
#if defined(_WIN32) || defined(_WIN64) // Windows
...
#define N_PATH_HOME  230
...
#else
...
#define N_PATH_HOME  160
...
#endif
```

160 bytes is too small to contain the 158 characters "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_audio_espeak-ng/espeak-ng/work/espeak-ng-1.50" plus the 15 characters "/espeak-ng-data" plus the terminating null.

So this raises many questions.

1. Why does espeak-ng need an absolute path to the espeak-ng-data directory? Could a relative path be used instead which would surely fit within 160 bytes?
2. Why does espeak-ng not detect when it is trying to create a string that is too long for the buffer and issue an error message? (`snprintf` has a return value which espeak-ng is ignoring.)
3. Why does Windows get a larger buffer size than other operating systems?
4. How were the sizes of 230 and 160 bytes selected? These values have already been increased several times over the years but without explanation. `N_PATH_HOME` was originally defined with a value of 120 in 810fb2404d0ba028144e12b6cd087b051febdc5d, replacing previously hardcoded values. It was increased from 120 to 220 on Windows and 150 on other platforms in f9725f9f64d672d94979fcd9531621dc0d709228 but the commit message doesn't say why. It was increased from 220 to 230 on Windows and from 150 to 160 on other platforms in 2d64d3289318b0e2b400ad3adb47aaf8726854a0 but the commit message doesn't say why.

**Comment by jaacoppi** [comment]
Thanks!

The original author is no longer part of the project so we'll probably never know why N_PATH_HOME was implemented like that. My guess is that it was just a quick and dirty hack.

This should be changed. I think your points 1 & 2 have the proper solution.

P.S.
You've commented on a lot of issues. It'll take some time to reply to them all.
-- 
Juho Hiltunen

**Comment by ryandesign** [comment]
> You've commented on a lot of issues. It'll take some time to reply to them all.

Mostly just looking to see if any existing issue covered what I was seeing or if there was anything I could provide an answer for, but also trying to point out issues that have already been answered and could be closed. Wading through over 300 open issues is time consuming so I recommend closing those that have already been dealt with.

**Comment by jaacoppi** [comment]
I'm assuming a relative path should work, but path_home is used in many places so it'll take time to check all the cases. At least the call to snprintf() should check the return value to see if there's an error. Another option would be to dynamically allocate memory for the path.

> > You've commented on a lot of issues. It'll take some time to reply to them all.
> 
> Mostly just looking to see if any existing issue covered what I was seeing or if there was anything I could provide an answer for, but also trying to point out issues that have already been answered and could be closed. Wading through over 300 open issues is time consuming so I recommend closing those that have already been dealt with.

Normally we expect the original authors to close their tickets if they feel that the issue has been solved for them. You've done good work researching and explaining why the tickets could be closed. I'll wait to see if there's any objections, then close them.

**Comment by ryandesign** [comment]
Just a reminder that it's been a year and this problem has not yet been fixed.

**Comment by beqabeqa473** [comment]
I suspect it is fixed in 3857b83c19af0bd4d4ea8f44f9e9baf6167f0837

**Comment by djphoenix** [comment]
Indeed, should be available to change with `-DNPATH_HOME=...` in CFLAGS during `configure`
